nomad/state/state_store.go

```diff
@@ -2405,17 +2405,17 @@ func (s *StateStore) UpsertCSIVolume(index uint64, volumes []*structs.CSIVolume)
 		}
 		if obj != nil {
 			// Allow some properties of a volume to be updated in place, but
-			// prevent accidentally overwriting important properties, or
-			// overwriting a volume in use
+			// prevent accidentally overwriting important properties.
 			old := obj.(*structs.CSIVolume)
 			if old.ExternalID != v.ExternalID ||
 				old.PluginID != v.PluginID ||
 				old.Provider != v.Provider {
 				return fmt.Errorf("volume identity cannot be updated: %s", v.ID)
 			}
-			s.CSIVolumeDenormalize(nil, old.Copy())
```

Comment:
`CSIVolumeDenormalize` is what makes sure the allocs are populated in the volume. Looks like we needed those for the `InUse` method call to verify we weren't changing something that was in use. Which is now ok so we can safely get rid of `CSIVolumeDenormalize`.

_But_ now that's making me realize this use of `MergeWhileInUse` doesn't make sense at all. We don't know that `old` is in use without checking `InUse`. And even if we did, the merge is mutating `old` which we never use again except to get the `CreateIndex`.

Reply:
> `CSIVolumeDenormalize` is what makes sure the allocs are populated in the volume.

Where is it actually populating something here?  I don't see anything persistent from this call to anywhere outside of it...

Part of `MergeWhileInUse` confusion might be a naming issue? What it does is "merge stuff that is safe to merge, even while in use" whether it actually is in use or not.

But you're totally right about `old` not really being used, that's a mistake.  The intention was to take the current ("old") volume, and only update it with the ok-to-change values, instead of trying to denylist things that can't change.  I missed the extra step though that reassigns `*v = *old`

Reply:
By "persistent" I mean I don't see any side-effects at all that reach outside of this function call.  Maybe I'm not following how `old.Copy()` works though?  Is denormalizing `old.Copy()` also affecting `old` itself?

Reply:
To clarify a bit, I've gathered (I think) that "denormalizing" is an in-memory thing to enrich data, usually across raft tables (in this case, csi_volumes and alloc tables), but here I'm not seeing how anything is being enriched outside of the call :shrug:

Reply:
_Oh._ I see. We're not returning the value by doing `old = s.CSIDenormalize(nil, old.Copy())`. That means `old.InUse()` is operating over the wrong object (the original old, and not the copy). That'd be alarming but it doesn't look like `InUse` cares about objects in the `PastClaims` field, and the length of those `ReadAllocs` and `WriteAllocs` doesn't change, so it doesn't result in a bug.

That's very strange. Looks like it the denormalize was added [here](https://github.com/hashicorp/nomad/commit/7d0f87b9106a8557896c6cc09576c846235495a5#diff-53c7d524897a192ece819816a9dcf006dee46747fbb853e7f638601ea556a72cR2172) but it never was useful to be doing. We probably should remove it from the backports then too, just so we don't do this work for nothing and leave this landmine of confusion in the codebase. :blush:

Reply:
Nice, phew, reassuring that I wasn't just missing something!

I'm open to some splitting-out for backporting separate PR(s), but not sure where to cleave the changes.  Maybe everything from Register/Create into reconcileVolume, but then exclude expandVolume and everything after that?  That could include the ["no healthy controllers"](https://github.com/hashicorp/nomad/pull/18359#discussion_r1318826936) change too.  Or, I could remove just this one line...?

Reply:
I would land this in main and then do a brand new PR that targets the most recent backport branch (`release/1.6.x`) with just the backportable fixes. And then once that merges to `release/1.6.x`, cherry-pick it to the other backport branches. That way you don't need to mess with this branch at all and it lands as once nice clean body of work in `main`.